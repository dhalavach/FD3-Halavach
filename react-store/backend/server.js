import express from 'express';
import cors from 'cors';
import { readFile } from 'fs/promises';

const app = express();
const port = 3000;

// Enable CORS for all routes
app.use(cors());

// Example product data
// const products = [
//   {
//     id: 1,
//     className: 'Item',
//     itemType: 'motherboard',
//     itemName: 'ASUS ROG STRIX B550-I GAMING',
//     itemPrice: 730,
//     itemImage: '/assets/images/asusb550i.jpg',
//     itemImageURL: 'https://uni.by/files/ItemsImages/116378_r300.jpg',
//     itemImageAlt: 'ASUS B550-I',
//     itemQuantity: 2,
//     itemDescription: '',
//   },
//   {
//     id: 2,
//     className: 'Item',
//     itemType: 'motherboard',
//     itemName: 'ASUS PRIME Z790-P D4',
//     itemPrice: 820,
//     itemImage: '/assets/images/asusz790.jpg',
//     itemImageURL: 'https://uni.by/files/ItemsImages/135354_r300.jpg',
//     itemImageAlt: 'Asus Z790 Prime',
//     itemQuantity: 1,
//     itemDescription: '',
//   },
//   {
//     id: 3,
//     className: 'Item',
//     itemType: 'motherboard',
//     itemName: 'MSI MPG B650 Edge WiFi',
//     itemPrice: 1095,
//     itemImage: '/assets/images/msib650.jpg',
//     itemImageURL: 'https://uni.by/files/ItemsImages/137010_r300.jpg',
//     itemImageAlt: 'MSI B650 Edge',
//     itemQuantity: 0,
//     itemDescription: '',
//   },
// ];
// const products = await readFile('./db.json', 'utf-8', (err, data) => {
//   if (err) {
//     console.log(err);
//   }
//   return data;
// });

// Define an endpoint to serve products
app.get('/products', async (req, res) => {
  try {
    // Reading products.json file asynchronously
    const data = await readFile(new URL('./db.json', import.meta.url), 'utf-8');
    const products = JSON.parse(data); // Parse the JSON data
    res.json(products);
  } catch (error) {
    console.error('Error reading db.json file', error);
    res.status(500).json({ message: 'Server error: unable to load products' });
  }
});
// Start the server
app.listen(port, () => {
  console.log(`Server is running on http://localhost:${port}`);
});
