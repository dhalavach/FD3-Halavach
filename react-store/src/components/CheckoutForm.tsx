import axios from 'axios';
import React, { useState, useEffect } from 'react';
import { Product } from '../types/Types';
import { useDispatch, useSelector } from 'react-redux';
import { setOrders } from '../slices/ordersSlice';

export default function CheckoutForm(props: any) {
  const [name, setName] = useState('');
  const [email, setEmail] = useState('');
  const [address, setAddress] = useState('');
  const [emailError, setEmailError] = useState('');
  const [nameError, setNameError] = useState('');
  const [addressError, setAddressError] = useState('');
  const [userInputPresent, setUserInputPresent] = useState({
    email: false,
    name: false,
    address: false,
  });
  const dispatch = useDispatch();
  const cartProducts = useSelector((state: any) => state.cartProducts);

  const postConfig = {
    URL: 'http://localhost:3000/orders',
    method: 'POST',
    headers: {
      Accept: 'application/json',
    },
  };

  const validateName = (name: string) => {
    const message =
      'Name must be longer or equal than 3 and shorter or equal than 100 characters!';
    if (!name) setNameError(message);
    else if (name.length >= 3 && name.length <= 100) {
      setNameError('');
    } else setNameError(message);
  };

  const validateEmail = (email: string) => {
    const message = 'Enter valid email address!';
    const emailRegex = new RegExp(
      /(?:[a-z0-9!#$%&'*+/=?^_`{|}~-]+(?:\.[a-z0-9!#$%&'*+/=?^_`{|}~-]+)*|"(?:[\x01-\x08\x0b\x0c\x0e-\x1f\x21\x23-\x5b\x5d-\x7f]|\\[\x01-\x09\x0b\x0c\x0e-\x7f])*")@(?:(?:[a-z0-9](?:[a-z0-9-]*[a-z0-9])?\.)+[a-z0-9](?:[a-z0-9-]*[a-z0-9])?|\[(?:(?:(2(5[0-5]|[0-4][0-9])|1[0-9][0-9]|[1-9]?[0-9]))\.){3}(?:(2(5[0-5]|[0-4][0-9])|1[0-9][0-9]|[1-9]?[0-9])|[a-z0-9-]*[a-z0-9]:(?:[\x01-\x08\x0b\x0c\x0e-\x1f\x21-\x5a\x53-\x7f]|\\[\x01-\x09\x0b\x0c\x0e-\x7f])+)\])/
    );
    if (!email) setEmailError(message);
    else if (email.match(emailRegex)) setEmailError('');
    else setEmailError(message);
  };

  const validateAddress = (address: string) => {
    const message = 'Enter valid shipping address!';
    const addressRegex = new RegExp(/[A-Za-z0-9'\.\-\s\,]/);
    if (!address) setAddressError(message);
    else if (address.match(addressRegex) && address.length >= 5)
      setAddressError('');
    else setAddressError(message);
  };

  const validateBeforeSubmit = (): boolean => {
    if (
      Object.values(userInputPresent).every((val) => val === true) &&
      !nameError &&
      !emailError &&
      !addressError
    ) {
      return true;
    } else {
      return false;
    }
  };

  const postOrder = (data: {
    name: string;
    email: string;
    address: string;
    orderedProducts: Product[];
  }) => {
    dispatch(setOrders(data));

    try {
      axios.post(postConfig.URL, data).then((response) => {
        console.log(response.status, response.data);
      });
    } catch (error) {
      console.error(error);
    }
  };

  return (
    <>
      <div className='cart'>CheckoutForm</div>
      <form>
        <ul className='form-container'>
          <li>
            <label>Email</label>
            <input
              name='email'
              type='email'
              required
              value={email}
              data-testid='checkout__email'
              onChange={(e) => {
                setEmail(e.target.value);
                validateEmail(e.target.value);
                setUserInputPresent({ ...userInputPresent, email: true });
              }}
            ></input>
            <span
              className='message-error'
              data-testid='checkout__email-error-message'
            >
              {emailError}
            </span>
          </li>
          <li>
            <label>First and last name</label>
            <input
              name='name'
              type='text'
              required
              value={name}
              data-testid='checkout__name'
              onChange={(e) => {
                setName(e.target.value);
                validateName(e.target.value);
                setUserInputPresent({ ...userInputPresent, name: true });
              }}
            ></input>
            <span
              className='message-error'
              data-testid='checkout__name-error-message'
            >
              {nameError}
            </span>
          </li>
          <li>
            <label>Delivery address</label>
            <input
              name='address'
              type='text'
              required
              value={address}
              data-testid='checkout__address'
              onChange={(e) => {
                setAddress(e.target.value);
                validateAddress(e.target.value);
                setUserInputPresent({ ...userInputPresent, address: true });
              }}
            ></input>
            <span
              className='message-error'
              data-testid='checkout__address-error-message'
            >
              {addressError}
            </span>
          </li>
          <li>
            <button
              className={`btn primary btn-submit ${
                validateBeforeSubmit() ? '' : 'grey-out'
              }`}
              type='submit'
              onClick={(e) => {
                e.preventDefault();
                if (validateBeforeSubmit()) {
                  postOrder({
                    name: name,
                    email: email,
                    address: address,
                    orderedProducts: cartProducts,
                  });
                }
              }}
              data-testid='checkout__submit-button'
            >
              Proceed to checkout
            </button>
          </li>
        </ul>
      </form>
    </>
  );
}
