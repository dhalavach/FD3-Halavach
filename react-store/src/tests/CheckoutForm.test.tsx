import { cleanup, fireEvent, render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import '@testing-library/jest-dom';

import CheckoutForm from '../components/CheckoutForm';
import { StaticRouter as Router } from 'react-router-dom/server';
import { Provider } from 'react-redux';
import store from '../components/Store';
import renderer from 'react-test-renderer';

afterEach(cleanup);

test('Renders the component', () => {
  const form = render(
    <Provider store={store}>
      <Router location={''}>
        <CheckoutForm setCheckoutFormOpen={() => {}} />
      </Router>
    </Provider>
  );
  expect(form).toBeTruthy();
});

test('submit button is greyed-out on load', () => {
  const form = render(
    <Provider store={store}>
      <Router location={''}>
        <CheckoutForm setCheckoutFormOpen={() => {}} />
      </Router>
    </Provider>
  );

  expect(
    form.getByTestId('checkout__submit-button').classList.contains('grey-out')
  ).toBe(true);
  form.unmount();
});

test('submit button is greyed-out with invalid email', () => {
  const form = render(
    <Provider store={store}>
      <Router location={''}>
        <CheckoutForm setCheckoutFormOpen={() => {}} />
      </Router>
    </Provider>
  );

  userEvent.type(form.getByTestId('checkout__email'), 'a');

  expect(
    form.getByTestId('checkout__submit-button').classList.contains('grey-out')
  ).toBe(true);
  form.unmount();
});

test('submit button is not greyed-out with valid input', async () => {
  const form = render(
    <Provider store={store}>
      <Router location={''}>
        <CheckoutForm setCheckoutFormOpen={() => {}} />
      </Router>
    </Provider>
  );
  const email = form.getByTestId('checkout__email');
  const name = form.getByTestId('checkout__name');
  const address = form.getByTestId('checkout__address');

  await userEvent.type(email, 'asdf@mail.com');
  await userEvent.type(name, 'John Doe');
  await userEvent.type(address, 'Some Street 1, SomeCity');

  expect(
    form.getByTestId('checkout__submit-button').classList.contains('grey-out')
  ).toBe(false);
  form.unmount();
});

test('submit button is greyed-out with invalid input', async () => {
  const form = render(
    <Provider store={store}>
      <Router location={''}>
        <CheckoutForm setCheckoutFormOpen={() => {}} />
      </Router>
    </Provider>
  );
  const email = form.getByTestId('checkout__email');
  const name = form.getByTestId('checkout__name');
  const address = form.getByTestId('checkout__address');
  const submitButton = form.getByTestId('checkout__submit-button');
  await userEvent.type(email, 'asdf');
  await userEvent.type(name, 'J');
  await userEvent.type(address, '3');
  expect(submitButton).toHaveClass('grey-out');
  form.unmount();
});

test('submit button renders with correct text', () => {
  const form = render(
    <Provider store={store}>
      <Router location={''}>
        <CheckoutForm setCheckoutFormOpen={() => {}} />
      </Router>
    </Provider>
  );
  const submitButton = form.getByTestId('checkout__submit-button');
  expect(submitButton.textContent).toBe('Proceed to checkout');
  form.unmount();
});

test('error message is displayed when invalid email is entered', () => {
  const form = render(
    <Provider store={store}>
      <Router location={''}>
        <CheckoutForm setCheckoutFormOpen={() => {}} />
      </Router>
    </Provider>
  );
  const emailInput = screen.getByTestId('checkout__email');
  fireEvent.change(emailInput, { target: { value: 'qwerty' } });
  const errorMessage = screen.getByTestId('checkout__email-error-message');
  expect(errorMessage).toBeVisible();
  form.unmount();
});

test('error message is displayed when invalid name is entered', () => {
  const form = render(
    <Provider store={store}>
      <Router location={''}>
        <CheckoutForm setCheckoutFormOpen={() => {}} />
      </Router>
    </Provider>
  );
  const nameInput = screen.getByTestId('checkout__name');
  fireEvent.change(nameInput, { target: { value: 'k' } });
  const errorMessage = screen.getByTestId('checkout__name-error-message');
  expect(errorMessage).toBeVisible();
  form.unmount();
});

test('error message is displayed when invalid address is entered', () => {
  const form = render(
    <Provider store={store}>
      <Router location={''}>
        <CheckoutForm setCheckoutFormOpen={() => {}} />
      </Router>
    </Provider>
  );
  const addressInput = screen.getByTestId('checkout__address');
  fireEvent.change(addressInput, { target: { value: 1 } });
  const errorMessage = screen.getByTestId('checkout__address-error-message');
  expect(errorMessage).toBeVisible();
  form.unmount();
});

test('component matches the snapshot', () => {
  const tree = renderer
    .create(
      <Provider store={store}>
        <Router location={''}>
          <CheckoutForm setCheckoutFormOpen={() => {}} />
        </Router>
      </Provider>
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});
