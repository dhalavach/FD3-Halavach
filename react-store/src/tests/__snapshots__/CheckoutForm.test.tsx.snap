// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`component matches the snapshot 1`] = `
[
  <div
    className="cart"
  >
    CheckoutForm
  </div>,
  <form>
    <ul
      className="form-container"
    >
      <li>
        <label>
          Email
        </label>
        <input
          data-testid="checkout__email"
          name="email"
          onChange={[Function]}
          required={true}
          type="email"
          value=""
        />
        <span
          className="message-error"
          data-testid="checkout__email-error-message"
        />
      </li>
      <li>
        <label>
          First and last name
        </label>
        <input
          data-testid="checkout__name"
          name="name"
          onChange={[Function]}
          required={true}
          type="text"
          value=""
        />
        <span
          className="message-error"
          data-testid="checkout__name-error-message"
        />
      </li>
      <li>
        <label>
          Delivery address
        </label>
        <input
          data-testid="checkout__address"
          name="address"
          onChange={[Function]}
          required={true}
          type="text"
          value=""
        />
        <span
          className="message-error"
          data-testid="checkout__address-error-message"
        />
      </li>
      <li>
        <button
          className="btn primary btn-submit grey-out"
          data-testid="checkout__submit-button"
          onClick={[Function]}
          type="submit"
        >
          Proceed to checkout
        </button>
      </li>
    </ul>
  </form>,
  <div />,
]
`;
